packages/conform-zod/v4/coercion.ts

```diff
@@ -0,0 +1,415 @@
+	// This is to prevent infinite recursion caused by z.lazy()
+	if (result) {
+		return result;
+	}
+
+	let schema: $ZodType = type;
+	const zod = (type as unknown as $ZodTypes)._zod;
+	const def = zod.def;
+	const constr = zod.constr;
+
+	const coercion = options.coerce(type);
+
+	if (coercion) {
+		schema = pipe(transform(coercion), type);
+	} else if (def.type === 'array') {
+		schema = pipe(
+			transform((value) => {
+				// No preprocess needed if the value is already an array
+				if (Array.isArray(value)) {
+					return value;
+				}
+
+				if (
+					typeof value === 'undefined' ||
+					typeof options.stripEmptyValue(value) === 'undefined'
+				) {
+					return [];
+				}
+
+				// Wrap it in an array otherwise
+				return [value];
+			}),
+			new constr({
+				...def,
+				element: enableTypeCoercion(def.element, options),
+			}),
+		);
+	} else if (def.type === 'object') {
+		schema = pipe(
+			transform((value) => {
+				if (typeof value === 'undefined') {
+					// Defaults it to an empty object
+					return {};
+				}
+
+				return value;
+			}),
+			new constr({
+				...def,
+				shape: Object.fromEntries(
+					Object.entries(def.shape).map(([key, def]) => [
+						key,
+						enableTypeCoercion(def, options),
+					]),
+				),
+			}),
+		);
+	} else if (def.type === 'optional') {
+		schema = pipe(
+			transform(options.stripEmptyValue),
+			new constr({
+				...def,
+				innerType: enableTypeCoercion(def.innerType, options),
+			}),
+		);
+	} else if (def.type === 'default') {
+		schema = pipe(
+			transform(options.stripEmptyValue),
+			new constr({
+				...def,
+				innerType: enableTypeCoercion(def.innerType, options),
+			}),
+		);
+	} else if (def.type === 'catch') {
+		schema = new constr({
+			...def,
+			innerType: enableTypeCoercion(def.innerType, options),
+		});
+	} else if (def.type === 'intersection') {
+		schema = new constr({
+			...def,
+			left: enableTypeCoercion(def.left, options),
+			right: enableTypeCoercion(def.right, options),
+		});
+	} else if (def.type === 'union') {
+		// From Zod v4, the `discriminatedUnion` schema is also defined as a `union`. I would like to use `constr`, i.e. `discriminatedUnion` like other schemas, but I can't do that.
+		// The reason is that `discriminatedUnion` cannot define the following schema.
+		// z.discriminatedUnion('type', [
+		//   z.any()
+		//     .transform(v => v == null ? {} : v)
+		//     .pipe(
+		//       z.object({
+		//         type: z.literal('a'),
+		//         number: z.number(),
+		//       })
+		//     ),
+		// ]);
+		// The `options` of `discriminatedUnion` basically defines the `object` schema, but `@conform-to/zod` requires `transform` to convert Object to an empty object when it is `undefined`, so the `union` schema is used.
+		// `discriminatedUnion` and `union` are strictly different schemas, but they behave similarly.
+		schema = union(
+			def.options.map((option) => enableTypeCoercion(option, options)),
+		);
```

Comment:
The example looks like zod 3? Is it still a problem now that we have v4 exported on a different module?

Reply:
This is a type error when defined in Zod v4. If you ignore the type and run it, an exception will be thrown and it will not run correctly.
![image](https://github.com/user-attachments/assets/b016c86e-542d-4db4-92be-a94351a26216)

Zod v3 is also slightly different, but the type is an error.
![image](https://github.com/user-attachments/assets/1e83a50d-e527-4790-bb54-3dee1fe90e4e)

It may be possible to avoid this by instantiating `ZodDiscriminatedUnion` as in Zod v3, but I haven't looked into it that far. My guess is that there may not be such a workaround in `@zod/mini`.

Reply:
Zod v4 has been modified to redefine `discriminatedUnion`. The logic has been modified to redefine it from the user-defined schema stored in Zod.